longturn/freeciv21
# Improve tileset error handling

This PR makes architectural changes to the way tileset errors are handled to avoid situations like #806. Errors and warnings are now saved in memory and displayed in the tileset debugger. When fixable errors are detected, a single dialog box is opened instead of one per error previously. The only recoverable issue is missing sprites for the terrain layers.

Here is a screenshot in an error situation (notice black tiles underneath the dialog):
![image](https://user-images.githubusercontent.com/22327575/148721876-2e3fafc2-7818-4fd6-8c86-24e79f71bb3e.png)

The tileset dialog has gained a new tab with the message log:
![image](https://user-images.githubusercontent.com/22327575/148721974-c978390f-3357-47e5-88a4-33df41de585f.png)

Icons are from the theme. They will only appear on Linux and I'm not even sure every theme will have them.

I still need to update its documentation, but maybe I can get help on this one?

Closes #806.

**Testing plan:**

You will first want to reproduce #806.
:warning: This may freeze your X server! If this happens, go to a new VT with Ctrl+Alt+F2 or similar, log in and do `killall -9 freeciv21-client`. Make sure that you know how to do it before you try! :warning: 

1. Open `data/amplio2/terrain2.spec` and comment out all sprite definitions.
2. Start a game using `amplio2`
3. If there's no freeze, you'll see 64 message boxes

Now you can try the same with the patch. Relevant error messages should appear in the debugger. You can also try to change `gfx = "amplio2/terrain2"` to a non-existing name; since the file contains only terrain sprites, this case is also supported as a non-fatal tileset error.

Finally, there's a whole class of edge cases with zoom, since this basically reloads the tileset. So you should also try that.
